cmd/kubeadm/app/cmd/phases/init/waitcontrolplane.go

```diff
@@ -119,3 +138,83 @@ func newControlPlaneWaiter(dryRun bool, timeout time.Duration, client clientset.
 
 	return apiclient.NewKubeWaiter(client, timeout, out), nil
 }
+
+type startupResult bool
+
+type startupProber struct {
+	runners []func(ch chan<- startupResult)
+}
+
+// getControlPlaneComponentStartupProber get all control plane component startup probes
+func getControlPlaneComponentStartupProber(cfg *kubeadm.InitConfiguration) startupProber {
+	specs := controlplane.GetStaticPodSpecs(&cfg.ClusterConfiguration, &cfg.LocalAPIEndpoint, nil)
+	prober := startupProber{}
+	for name, podSpec := range specs {
+		for _, container := range podSpec.Spec.Containers {
+			p := container.StartupProbe
+			// Currently control plane component only use http probe
+			if p.HTTPGet != nil {
+				url := &url.URL{
+					Host:   net.JoinHostPort(p.HTTPGet.Host, p.HTTPGet.Port.String()),
+					Scheme: string(p.HTTPGet.Scheme),
+					Path:   p.HTTPGet.Path,
+				}
+				req, err := http.NewRequest("GET", url.String(), nil)
+				if err != nil {
+					klog.V(1).ErrorS(err, "[wait-control-plane] Failed to create control plane component startup probe request", "component", name)
+					continue
+				}
+
+				timeout := time.Duration(p.TimeoutSeconds) * time.Second
+				client := &http.Client{
+					Timeout: timeout,
+					Transport: &http.Transport{
+						TLSClientConfig:   &tls.Config{InsecureSkipVerify: true},
+						DisableKeepAlives: true,
+					},
+				}
+
+				runner := func(client *http.Client, req *http.Request, component string) func(chan<- startupResult) {
+					return func(ch chan<- startupResult) {
+						errMsg := "[wait-control-plane] Failed to startup control plane component"
+						res, err := client.Do(req)
+						if err != nil {
+							klog.V(1).ErrorS(err, errMsg, "component", component, "probe", req.URL)
+							ch <- false
+							return
+						}
+						defer res.Body.Close()
+
+						if res.StatusCode >= http.StatusOK && res.StatusCode < http.StatusMultipleChoices {
+							ch <- true
+						} else {
+							klog.V(1).ErrorS(nil, errMsg, "component", component, "probe", req.URL, "statuscode", res.StatusCode)
+							ch <- false
+						}
+					}
+				}(client, req, name)
+
+				prober.runners = append(prober.runners, runner)
+			}
+		}
+	}
+
+	return prober
+}
+
+// isStartup returns true if all startup probes success
+func (p *startupProber) isStartup() bool {
```

Comment:
```suggestion
func (p *startupProber) hasStartedUp() bool {
```

^ might be the more accurate naming